src/coreclr/jit/valuenum.cpp

```diff
@@ -8774,6 +8788,57 @@ void Compiler::fgValueNumberTree(GenTree* tree)
         }
         else // Look up the VNFunc for the node
         {
+            // Check if we can fold GT_ARR_LENGTH on top of a known array (immutable)
+            if (tree->OperIs(GT_ARR_LENGTH))
+            {
+                // Case 1: ARR_LENGTH(FROZEN_OBJ)
+                ValueNum addressVN = tree->gtGetOp1()->gtVNPair.GetLiberal();
+                if (vnStore->IsVNHandle(addressVN) && (vnStore->GetHandleFlags(addressVN) == GTF_ICON_OBJ_HDL))
+                {
+                    size_t handle = vnStore->CoercedConstantValue<size_t>(addressVN);
+                    int    len    = this->info.compCompHnd->getArrayOrStringLength((CORINFO_OBJECT_HANDLE)handle);
+                    if (len >= 0)
+                    {
+                        tree->gtVNPair.SetBoth(vnStore->VNForIntCon(len));
```

Comment:
It is not correct to skip adding exception sets. Imagine:
```
ARR_LENGTH
    COMMA
       <side effects>
       CNS_INT <frozen object>
```
This problem would be avoided if we move this folding logic into `VNForFunc(unary)`.